lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -131,6 +132,93 @@ static bool checkAsyncHandler(FuncDecl *func, bool diagnose) {
   return false;
 }
 
+/// Check whether the function is a proper distributed function
+///
+/// \param diagnose Whether to emit a diagnostic when a problem is encountered.
+///
+/// \returns \c true if there was a problem with adding the attribute, \c false
+/// otherwise.
+static bool checkDistributedFunc(FuncDecl *func, bool diagnose) {
+  // === All parameters and the result type must be Codable
+
+  auto &C = func->getASTContext();
+  auto encodableType = C.getProtocol(KnownProtocolKind::Encodable);
+  auto decodableType = C.getProtocol(KnownProtocolKind::Decodable);
+
+  // --- Check parameters for 'Codable' conformance
+  for (auto param : *func->getParameters()) {
+    auto paramType = func->mapTypeIntoContext(param->getInterfaceType()); // TODO: getDeclaredInterfaceType instead?
+    if (TypeChecker::conformsToProtocol(paramType, encodableType, func).isInvalid() ||
+        TypeChecker::conformsToProtocol(paramType, decodableType, func).isInvalid()) {
+      if (diagnose)
+        func->diagnose(
+            diag::distributed_actor_func_param_not_codable,
+            param->getArgumentName().str(),
+            param->getInterfaceType()
+        );
+      // TODO: suggest a fixit to add Codable to the type?
+      return true;
+    }
+  }
+
+  // --- Result type must be either void or a codable type
+  // TODO: In the future we can also support AsyncSequence of Codable values
```

Comment:
I thought a bit more about this one and I wonder what we'll do here...

This isn't quite right that the stream is Codable "per se" but we want to treat it as a reference to the local stream, and if subscribed to removely, it'd want to initiate streaming over this... So it's like an "actor ref" (in akka i called those "stream ref"), it may be weird to make `AsyncSequence` distributed conformance be aware of distributed actors, so rather we might want some other wrapper type... 🤔 

The TODO I can remove though, this would be future feature work, rather than something that's "missing"